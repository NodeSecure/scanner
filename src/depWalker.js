// Import Node.js Dependencies
import path from "path";
import { readFileSync, promises as fs } from "fs";
import timers from "timers/promises";
import os from "os";

// Import Third-party Dependencies
import combineAsyncIterators from "combine-async-iterators";
import iter from "itertools";
import pacote from "pacote";
import Arborist from "@npmcli/arborist";
import Lock from "@slimio/lock";
import * as vuln from "@nodesecure/vuln";
import { ScannerLoggerEvents } from "./constants.js";

// Import Internal Dependencies
import {
  mergeDependencies, getCleanDependencyName, getDependenciesWarnings, addMissingVersionFlags, isGitDependency,
  NPM_TOKEN
} from "./utils/index.js";
import { scanDirOrArchive } from "./tarball.js";
import { packageMetadata } from "./npmRegistry.js";
import Dependency from "./class/dependency.class.js";
import Logger from "./class/logger.class.js";

const { version: packageVersion } = JSON.parse(
  readFileSync(
    new URL(path.join("..", "package.json"), import.meta.url)
  )
);

export async function* searchDeepDependencies(packageName, gitURL, options) {
  const { exclude, currDepth = 0, parent, maxDepth, registry } = options;

  const { name, version, deprecated, ...pkg } = await pacote.manifest(gitURL ?? packageName, {
    ...NPM_TOKEN,
    registry,
    cache: `${os.homedir()}/.npm`
  });
  const { dependencies, customResolvers, alias } = mergeDependencies(pkg);

  const current = new Dependency(name, version, parent);
  current.alias = Object.fromEntries(alias);

  if (gitURL !== null) {
    current.isGit(gitURL);
    try {
      await pacote.manifest(`${name}@${version}`, {
        ...NPM_TOKEN,
        registry,
        cache: `${os.homedir()}/.npm`
      });
    }
    catch {
      current.existOnRemoteRegistry = false;
    }
  }
  current.addFlag("isDeprecated", deprecated === true);
  current.addFlag("hasCustomResolver", customResolvers.size > 0);
  current.addFlag("hasDependencies", dependencies.size > 0);

  if (currDepth !== maxDepth) {
    const config = {
      exclude, currDepth: currDepth + 1, parent: current, maxDepth, registry
    };

    const gitDependencies = iter.filter(customResolvers.entries(), ([, valueStr]) => isGitDependency(valueStr));
    for (const [depName, valueStr] of gitDependencies) {
      yield* searchDeepDependencies(depName, valueStr, config);
    }

    const depsNames = await Promise.all(iter.map(dependencies.entries(), getCleanDependencyName));
    for (const [fullName, cleanName, isLatest] of depsNames) {
      if (!isLatest) {
        current.addFlag("hasOutdatedDependency");
      }

      if (exclude.has(cleanName)) {
        exclude.get(cleanName).add(current.fullName);
      }
      else {
        exclude.set(cleanName, new Set([current.fullName]));
        yield* searchDeepDependencies(fullName, null, config);
      }
    }
  }

  yield current;
}

export async function* deepReadEdges(currentPackageName, options) {
  const { to, parent, exclude, fullLockMode, includeDevDeps, registry } = options;
  const { version, integrity = to.integrity } = to.package;

  const updatedVersion = version === "*" || typeof version === "undefined" ? "latest" : version;
  const current = new Dependency(currentPackageName, updatedVersion, parent);
  current.dev = to.dev;

  if (fullLockMode && !includeDevDeps) {
    const { deprecated, _integrity, ...pkg } = await pacote.manifest(`${currentPackageName}@${updatedVersion}`, {
      ...NPM_TOKEN,
      registry,
      cache: `${os.homedir()}/.npm`
    });
    const { customResolvers, alias } = mergeDependencies(pkg);

    current.alias = Object.fromEntries(alias);
    current.addFlag("hasValidIntegrity", _integrity === integrity);
    current.addFlag("isDeprecated");
    current.addFlag("hasCustomResolver", customResolvers.size > 0);

    if (isGitDependency(to.resolved)) {
      current.isGit(to.resolved);
    }
  }
  current.addFlag("hasDependencies", to.edgesOut.size > 0);

  for (const [packageName, { to: toNode }] of to.edgesOut) {
    if (toNode === null || (!includeDevDeps && toNode.dev)) {
      continue;
    }
    const cleanName = `${packageName}@${toNode.package.version}`;

    if (exclude.has(cleanName)) {
      exclude.get(cleanName).add(current.fullName);
    }
    else {
      exclude.set(cleanName, new Set([current.fullName]));
      yield* deepReadEdges(packageName, { parent: current, to: toNode, exclude, registry });
    }
  }
  yield current;
}

export async function* getRootDependencies(manifest, options) {
  const {
    maxDepth = 4, exclude,
    usePackageLock, fullLockMode, includeDevDeps,
    location,
    registry
  } = options;

  const { dependencies, customResolvers, alias } = mergeDependencies(manifest, void 0);
  const parent = new Dependency(manifest.name, manifest.version);
  parent.alias = Object.fromEntries(alias);

  try {
    await pacote.manifest(`${manifest.name}@${manifest.version}`, {
      ...NPM_TOKEN,
      registry,
      cache: `${os.homedir()}/.npm`
    });
  }
  catch {
    parent.existOnRemoteRegistry = false;
  }
  parent.addFlag("hasCustomResolver", customResolvers.size > 0);
  parent.addFlag("hasDependencies", dependencies.size > 0);

  let iterators;
  if (usePackageLock) {
    const arb = new Arborist({
      ...NPM_TOKEN,
      path: location,
      registry
    });
    let tree;
    try {
      await fs.access(path.join(location, "node_modules"));
      tree = await arb.loadActual();
    }
    catch {
      tree = await arb.loadVirtual();
    }

    iterators = [
      ...iter
        .filter(tree.edgesOut.entries(), ([, { to }]) => to !== null && (includeDevDeps ? true : (!to.dev || to.isWorkspace)))
        .map(([packageName, { to }]) => [packageName, to.isWorkspace ? to.target : to])
        .map(([packageName, to]) => deepReadEdges(packageName, {
          to,
          parent,
          fullLockMode,
          includeDevDeps,
          exclude,
          registry
        }))
    ];
  }
  else {
    const configRef = { exclude, maxDepth, parent, registry };
    iterators = [
      ...iter.filter(customResolvers.entries(), ([, valueStr]) => isGitDependency(valueStr))
        .map(([depName, valueStr]) => searchDeepDependencies(depName, valueStr, configRef)),
      ...iter.map(dependencies.entries(), ([name, ver]) => searchDeepDependencies(`${name}@${ver}`, null, configRef))
    ];
  }
  for await (const dep of combineAsyncIterators({}, ...iterators)) {
    yield dep;
  }

  // Add root dependencies to the exclude Map (because the parent is not handled by searchDeepDependencies)
  // if we skip this the code will fail to re-link properly dependencies in the following steps
  const depsName = await Promise.all(iter.map(dependencies.entries(), getCleanDependencyName));
  for (const [, fullRange, isLatest] of depsName) {
    if (!isLatest) {
      parent.addFlag("hasOutdatedDependency");
    }
    if (exclude.has(fullRange)) {
      exclude.get(fullRange).add(parent.fullName);
    }
  }

  yield parent;
}

/**
 * @param {*} manifest
 * @param {*} options
 * @param {Logger} logger
 */
export async function depWalker(manifest, options = {}, logger = new Logger()) {
  const {
    forceRootAnalysis = false,
    usePackageLock = false,
    includeDevDeps = false,
    fullLockMode = false,
    maxDepth,
    location,
    vulnerabilityStrategy = vuln.strategies.NONE,
    registry
  } = options;

  // Create TMP directory
  const tmpLocation = await fs.mkdtemp(path.join(os.tmpdir(), "/"));

  const payload = {
    id: tmpLocation.slice(-6),
    rootDependencyName: manifest.name,
    scannerVersion: packageVersion,
    vulnerabilityStrategy,
    warnings: []
  };

  // We are dealing with an exclude Map to avoid checking a package more than one time in searchDeepDependencies
  const exclude = new Map();
  const dependencies = new Map();

  {
    logger
      .start(ScannerLoggerEvents.analysis.tree)
      .start(ScannerLoggerEvents.analysis.tarball)
      .start(ScannerLoggerEvents.analysis.registry);
    const fetchedMetadataPackages = new Set();
    const promisesToWait = [];

    const tarballLocker = new Lock({ maxConcurrent: 5 });
    tarballLocker.on("freeOne", () => logger.tick(ScannerLoggerEvents.analysis.tarball));

    const rootDepsOptions = { maxDepth, exclude, usePackageLock, fullLockMode, includeDevDeps, location, registry };
    for await (const currentDep of getRootDependencies(manifest, rootDepsOptions)) {
      const { name, version, dev } = currentDep;
      const current = currentDep.exportAsPlainObject(name === manifest.name ? 0 : void 0);
      let proceedDependencyAnalysis = true;

      if (dependencies.has(name)) {
        // TODO: how to handle different metadata ?
        const dep = dependencies.get(name);

        const currVersion = Object.keys(current.versions)[0];
        if (currVersion in dep.versions) {
          // The dependency has already entered the analysis
          // This happens if the package is used by multiple packages in the tree
          proceedDependencyAnalysis = false;
        }
        else {
          dep.versions[currVersion] = current.versions[currVersion];
        }
      }
      else {
        dependencies.set(name, current);
      }

      // If the dependency is a DevDependencies we ignore it.
      if (dev) {
        continue;
      }

      if (proceedDependencyAnalysis) {
        logger.tick(ScannerLoggerEvents.analysis.tree);

        // There is no need to fetch 'N' times the npm metadata for the same package.
        if (fetchedMetadataPackages.has(name)) {
          logger.tick(ScannerLoggerEvents.analysis.registry);
        }
        else {
          fetchedMetadataPackages.add(name);
          promisesToWait.push(packageMetadata(name, version, {
            ref: current,
            logger
          }));
        }

        promisesToWait.push(scanDirOrArchive(name, version, {
          ref: current.versions[version],
          location,
          tmpLocation: forceRootAnalysis && name === manifest.name ? null : tmpLocation,
          locker: tarballLocker,
          logger,
          registry
        }));
      }
    }

    logger.end(ScannerLoggerEvents.analysis.tree);

    // Wait for all extraction to be done!
    await Promise.allSettled(promisesToWait);
    await timers.setImmediate();

    logger.end(ScannerLoggerEvents.analysis.tarball).end(ScannerLoggerEvents.analysis.registry);
  }

  const { hydratePayloadDependencies, strategy } = await vuln.setStrategy(vulnerabilityStrategy);
  await hydratePayloadDependencies(dependencies, {
    useStandardFormat: true,
    path: location
  });

  payload.vulnerabilityStrategy = strategy;

  // We do this because it "seem" impossible to link all dependencies in the first walk.
  // Because we are dealing with package only one time it may happen sometimes.
  for (const [packageName, dependency] of dependencies) {
    for (const [verStr, verDescriptor] of Object.entries(dependency.versions)) {
      verDescriptor.flags.push(...addMissingVersionFlags(new Set(verDescriptor.flags), dependency));

      const fullName = `${packageName}@${verStr}`;
      const usedDeps = exclude.get(fullName) || new Set();
      if (usedDeps.size === 0) {
        continue;
      }

      const usedBy = Object.create(null);
      for (const [name, version] of [...usedDeps].map((name) => name.split(" "))) {
        usedBy[name] = version;
      }
      Object.assign(verDescriptor.usedBy, usedBy);
    }
  }

  try {
    const { warnings, authors } = await getDependenciesWarnings(dependencies);
    payload.warnings = warnings;
    payload.authors = authors;
    payload.dependencies = Object.fromEntries(dependencies);

    return payload;
  }
  finally {
    await timers.setImmediate();
    await fs.rm(tmpLocation, { recursive: true, force: true });

    logger.emit(ScannerLoggerEvents.done);
  }
}
