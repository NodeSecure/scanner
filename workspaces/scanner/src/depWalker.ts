// Import Node.js Dependencies
import path from "node:path";
import { readFileSync, promises as fs } from "node:fs";
import timers from "node:timers/promises";
import os from "node:os";

// Import Third-party Dependencies
import { Mutex, MutexRelease } from "@openally/mutex";
import { scanDirOrArchive, type scanDirOrArchiveOptions } from "@nodesecure/tarball";
import * as vuln from "@nodesecure/vuln";
import * as treeWalker from "@nodesecure/tree-walker";
import type { ManifestVersion, PackageJSON } from "@nodesecure/npm-types";

// Import Internal Dependencies
import {
  getDependenciesWarnings, addMissingVersionFlags
} from "./utils/index.js";
import { packageMetadata, manifestMetadata } from "./npmRegistry.js";
import { Logger, ScannerLoggerEvents } from "./class/logger.class.js";
import type {
  Dependency,
  DependencyVersion,
  Options,
  Payload
} from "./types.js";

// CONSTANTS
const kDefaultDependencyVersionFields = {
  description: "",
  size: 0,
  author: null,
  engines: {},
  scripts: {},
  license: [],
  composition: {
    extensions: [],
    files: [],
    minified: [],
    unused: [],
    missing: [],
    required_files: [],
    required_nodejs: [],
    required_thirdparty: [],
    required_subpath: []
  }
};
const kDefaultDependencyMetadata: Dependency["metadata"] = {
  publishedCount: 0,
  lastUpdateAt: new Date(),
  lastVersion: "N/A",
  hasChangedAuthor: false,
  hasManyPublishers: false,
  hasReceivedUpdateInOneYear: true,
  homepage: null,
  author: null,
  publishers: [],
  maintainers: [],
  integrity: {}
};

const { version: packageVersion } = JSON.parse(
  readFileSync(
    new URL(path.join("..", "package.json"), import.meta.url),
    "utf-8"
  )
);

type WalkerOptions = Omit<Options, "registry"> & {
  registry: string;
  location?: string;
}

export async function depWalker(
  manifest: PackageJSON | ManifestVersion,
  options: WalkerOptions,
  logger = new Logger()
): Promise<Payload> {
  const {
    scanRootNode = false,
    includeDevDeps = false,
    packageLock,
    maxDepth,
    location,
    vulnerabilityStrategy = vuln.strategies.NONE,
    registry
  } = options;

  // Create TMP directory
  const tmpLocation = await fs.mkdtemp(path.join(os.tmpdir(), "/"));

  const payload: Partial<Payload> = {
    id: tmpLocation.slice(-6),
    rootDependencyName: manifest.name,
    scannerVersion: packageVersion,
    vulnerabilityStrategy,
    warnings: []
  };

  // We are dealing with an exclude Map to avoid checking a package more than one time in searchDeepDependencies
  const exclude: Map<string, Set<string>> = new Map();
  const dependencies: Map<string, Dependency> = new Map();

  {
    logger
      .start(ScannerLoggerEvents.analysis.tree)
      .start(ScannerLoggerEvents.analysis.tarball)
      .start(ScannerLoggerEvents.analysis.registry);
    const fetchedMetadataPackages = new Set<string>();
    const promisesToWait: Promise<void>[] = [];

    const locker = new Mutex({ concurrency: 5 });
    locker.on(
      MutexRelease,
      () => logger.tick(ScannerLoggerEvents.analysis.tarball)
    );

    const rootDepsOptions: treeWalker.npm.WalkOptions = {
      maxDepth,
      exclude,
      includeDevDeps,
      registry,
      packageLock
    };
    for await (const current of treeWalker.npm.walk(manifest, rootDepsOptions)) {
      const { name, version, ...currentVersion } = current;
      const dependency: Dependency = {
        versions: {
          [version]: {
            ...currentVersion,
            ...structuredClone(kDefaultDependencyVersionFields)
          }
        },
        vulnerabilities: [],
        metadata: structuredClone(kDefaultDependencyMetadata)
      };

      let proceedDependencyScan = true;
      if (dependencies.has(name)) {
        const dep = dependencies.get(name)!;
        promisesToWait.push(
          manifestMetadata(name, version, dep)
        );

        if (version in dep.versions) {
          // The dependency has already entered the analysis
          // This happens if the package is used by multiple packages in the tree
          proceedDependencyScan = false;
        }
        else {
          dep.versions[version] = dependency.versions[version];
        }
      }
      else {
        dependencies.set(name, dependency);
      }

      // If the dependency is a DevDependencies we ignore it.
      if (current.isDevDependency || !proceedDependencyScan) {
        continue;
      }

      logger.tick(ScannerLoggerEvents.analysis.tree);

      // There is no need to fetch 'N' times the npm metadata for the same package.
      if (fetchedMetadataPackages.has(name) || !current.existOnRemoteRegistry) {
        logger.tick(ScannerLoggerEvents.analysis.registry);
      }
      else {
        fetchedMetadataPackages.add(name);
        promisesToWait.push(packageMetadata(name, version, {
          dependency,
          logger
        }));
      }

      const scanDirOptions = {
        ref: dependency.versions[version] as any,
        location,
        tmpLocation: scanRootNode && name === manifest.name ? null : tmpLocation,
        registry
      };
      promisesToWait.push(scanDirOrArchiveEx(name, version, locker, scanDirOptions));
    }

    logger.end(ScannerLoggerEvents.analysis.tree);

    // Wait for all extraction to be done!
    await Promise.allSettled(promisesToWait);
    await timers.setImmediate();

    logger
      .end(ScannerLoggerEvents.analysis.tarball)
      .end(ScannerLoggerEvents.analysis.registry);
  }

  const { hydratePayloadDependencies, strategy } = await vuln.setStrategy(vulnerabilityStrategy);
  await hydratePayloadDependencies(dependencies as any, {
    useStandardFormat: true,
    path: location
  });

  payload.vulnerabilityStrategy = strategy;

  // We do this because it "seem" impossible to link all dependencies in the first walk.
  // Because we are dealing with package only one time it may happen sometimes.
  const globalWarnings: string[] = [];
  for (const [packageName, dependency] of dependencies) {
    const metadataIntegrities = dependency.metadata?.integrity ?? {};

    for (const [version, integrity] of Object.entries(metadataIntegrities)) {
      const dependencyVer = dependency.versions[version] as DependencyVersion;

      // @ts-ignore
      const isEmptyPackage = dependencyVer.warnings.some((warning) => warning.kind === "empty-package");
      if (isEmptyPackage) {
        globalWarnings.push(`${packageName}@${version} only contain a package.json file!`);
      }

      if (!("integrity" in dependencyVer) || dependencyVer.flags.includes("isGit")) {
        continue;
      }

      if (dependencyVer.integrity !== integrity) {
        globalWarnings.push(`${packageName}@${version} manifest & tarball integrity doesn't match!`);
      }
    }
    for (const version of Object.entries(dependency.versions)) {
      const [verStr, verDescriptor] = version as [string, DependencyVersion];
      verDescriptor.flags.push(
        ...addMissingVersionFlags(new Set(verDescriptor.flags), dependency)
      );

      const usedDeps = exclude.get(`${packageName}@${verStr}`) || new Set();
      if (usedDeps.size === 0) {
        continue;
      }

      const usedBy: Record<string, string> = Object.create(null);
      for (const [name, version] of [...usedDeps].map((name) => name.split(" ") as [string, string])) {
        usedBy[name] = version;
      }
      Object.assign(verDescriptor.usedBy, usedBy);
    }
  }

  try {
    const { warnings, flaggedAuthors } = await getDependenciesWarnings(dependencies);
    payload.warnings = globalWarnings.concat(warnings);
    payload.flaggedAuthors = flaggedAuthors;
    payload.dependencies = Object.fromEntries(dependencies);

    return payload as Payload;
  }
  finally {
    await timers.setImmediate();
    await fs.rm(tmpLocation, { recursive: true, force: true });

    logger.emit(ScannerLoggerEvents.done);
  }
}

async function scanDirOrArchiveEx(
  name: string,
  version: string,
  locker: Mutex,
  options: scanDirOrArchiveOptions
) {
  const free = await locker.acquire();

  try {
    await scanDirOrArchive(name, version, options);
  }
  catch {
    // ignore
  }
  finally {
    free();
  }
}
