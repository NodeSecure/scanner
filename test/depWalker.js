// Require Node.js Dependencies
import { join } from "path";
import { readFileSync } from "fs";

// Third party Dependencies
import { setStrategy, strategies } from "@nodesecure/vuln";

// Require Internal Dependencies
import { depWalker } from "../src/depWalker.js";
import { from } from "../index.js";

// CONSTANTS
const FIXTURE_PATH = join("fixtures", "depWalker");

// JSON PAYLOADS
const is = JSON.parse(readFileSync(
  new URL(join(FIXTURE_PATH, "slimio.is.json"), import.meta.url)
));

const config = JSON.parse(readFileSync(
  new URL(join(FIXTURE_PATH, "slimio.config.json"), import.meta.url)
));

function cleanupPayload(payload) {
  for (const pkg of Object.values(payload)) {
    for (const versionName of pkg.versions) {
      pkg[versionName].composition.extensions.sort();
      delete pkg[versionName].size;
      delete pkg[versionName].composition.files;
      delete pkg[versionName].composition.required_files;
    }
  }
}

test("execute depWalker on @slimio/is", async() => {
  await setStrategy(strategies.NPM_AUDIT);

  const result = await depWalker(is, { verbose: false });
  const resultAsJSON = JSON.parse(JSON.stringify(result.dependencies, null, 2));
  cleanupPayload(resultAsJSON);

  expect(resultAsJSON).toMatchSnapshot();
});

test("execute depWalker on @slimio/config", async() => {
  await setStrategy(strategies.NPM_AUDIT);

  const result = await depWalker(config, { verbose: false });
  const resultAsJSON = JSON.parse(JSON.stringify(result.dependencies, null, 2));

  const packages = Object.keys(resultAsJSON).sort();
  expect(packages).toEqual([
    "lodash.clonedeep",
    "zen-observable",
    "lodash.set",
    "lodash.get",
    "node-watch",
    "fast-deep-equal",
    "fast-json-stable-stringify",
    "json-schema-traverse",
    "punycode",
    "uri-js",
    "ajv",
    "@slimio/is",
    "@iarna/toml",
    "@slimio/config"
  ].sort());
});

test("fetch payload of pacote on the npm registry", async() => {
  const result = await from("pacote", {
    verbose: false,
    maxDepth: 10,
    vulnerabilityStrategy: strategies.NPM_AUDIT
  });

  expect(Object.keys(result)).toMatchSnapshot();
});

test("fetch payload of pacote on the gitlat registry", async() => {
  const result = await from("pacote", {
    registry: "https://gitlab.com/api/v4/packages/npm/",
    verbose: false,
    maxDepth: 10,
    vulnerabilityStrategy: strategies.NPM_AUDIT
  });

  expect(Object.keys(result)).toMatchSnapshot();
});

test("Given an invalid URL, it should throw", async() => {
  expect.assertions(1);

  try {
    await from("pacote", {
      registry: "fake-url",
      verbose: false,
      maxDepth: 10,
      vulnerabilityStrategy: strategies.NPM_AUDIT
    });
  }
  catch (error) {
    expect(error.message).toBe("Invalid URL");
  }
});
