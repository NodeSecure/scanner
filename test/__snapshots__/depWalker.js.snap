// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute depWalker on @slimio/is 1`] = `
Object {
  "@slimio/is": Object {
    "1.5.1": Object {
      "author": "SlimIO",
      "composition": Object {
        "extensions": Array [
          "",
          ".js",
          ".json",
          ".md",
          ".toml",
          ".ts",
        ],
        "minified": Array [],
        "missing": Array [],
        "required_nodejs": Array [],
        "required_thirdparty": Array [],
        "unused": Array [],
      },
      "description": "SlimIO is (JavaScript Primitives &amp; Objects type checker)",
      "flags": Array [],
      "gitUrl": null,
      "id": 0,
      "license": Object {
        "hasMultipleLicenses": false,
        "licenses": Array [
          Object {
            "from": "package.json",
            "spdx": Object {
              "fsf": true,
              "fsfAndOsi": true,
              "includesDeprecated": false,
              "osi": true,
            },
            "spdxLicenseLinks": Array [
              "https://spdx.org/licenses/MIT.html#licenseText",
            ],
            "uniqueLicenseIds": Array [
              "MIT",
            ],
          },
          Object {
            "from": "LICENSE",
            "spdx": Object {
              "fsf": true,
              "fsfAndOsi": true,
              "includesDeprecated": false,
              "osi": true,
            },
            "spdxLicenseLinks": Array [
              "https://spdx.org/licenses/MIT.html#licenseText",
            ],
            "uniqueLicenseIds": Array [
              "MIT",
            ],
          },
        ],
        "uniqueLicenseIds": Array [
          "MIT",
        ],
      },
      "usedBy": Object {},
      "warnings": Array [],
    },
    "metadata": Object {
      "author": "SlimIO",
      "dependencyCount": 0,
      "hasManyPublishers": true,
      "hasReceivedUpdateInOneYear": false,
      "homepage": "https://github.com/SlimIO/is#readme",
      "lastUpdateAt": "2019-06-11T06:41:41.590Z",
      "lastVersion": "1.5.1",
      "maintainers": Array [
        Object {
          "email": "mickael.croquet@laposte.net",
          "name": "mcroquet",
        },
        Object {
          "email": "sebastien48criquet@gmail.com",
          "name": "ahkrin",
        },
        Object {
          "email": "gentilhomme.thomas@gmail.com",
          "name": "fraxken",
        },
        Object {
          "email": "alexandre.malaj@gmail.com",
          "name": "alexandre.malaj",
        },
      ],
      "publishedCount": 7,
      "publishers": Array [
        Object {
          "at": "2018-09-22T20:36:27.523Z",
          "name": "fraxken",
          "version": "1.0.0",
        },
      ],
    },
    "versions": Array [
      "1.5.1",
    ],
    "vulnerabilities": Array [],
  },
}
`;

exports[`fetch payload of pacote on the gitlat registry 1`] = `
Array [
  "id",
  "rootDepencyName",
  "warnings",
  "dependencies",
  "version",
  "vulnerabilityStrategy",
]
`;

exports[`fetch payload of pacote on the npm registry 1`] = `
Array [
  "id",
  "rootDepencyName",
  "warnings",
  "dependencies",
  "version",
  "vulnerabilityStrategy",
]
`;
